src/OrleansRuntime/Storage/MemoryStorageGrain.cs

```diff
@@ -66,57 +63,67 @@ private GrainStateStore GetStoreForGrain(string grainType)
+            public IGrainState GetGrainState(string grainId)
             {
-                Tuple<IDictionary<string, object>, long> state;
-                if(grainStateStorage.TryGetValue(grainId, out state))
-                    return Tuple.Create(state.Item1, state.Item2.ToString());
-                else
-                    return Tuple.Create<IDictionary<string, object>, string>(null, null); // upon first read, return null/invalid etag, to mimic Azure Storage.
+                IGrainState entry;
+                grainStateStorage.TryGetValue(grainId, out entry);
+                return entry;
             }
 
-            public string UpdateGrainState(string grainStoreKey, IDictionary<string, object> state, string receivedEtag)
+            public string UpdateGrainState(string grainId, IGrainState grainState)
             {
-                long currentETag = 0;
-                Tuple<IDictionary<string, object>, long> oldState;
-                if (grainStateStorage.TryGetValue(grainStoreKey, out oldState)) {
-                    currentETag = oldState.Item2;
+                IGrainState entry;
+                grainStateStorage.TryGetValue(grainId, out entry);
+                if (entry == null)
+                {
+                    grainStateStorage[grainId] = grainState;
+                    return grainState.ETag;
                 }
-                ValidateEtag(currentETag, receivedEtag, grainStoreKey, "Update");
-                currentETag++;
-                grainStateStorage[grainStoreKey] = Tuple.Create(state, currentETag);
-                return currentETag.ToString();
+
+                ValidateEtag(grainState.ETag, entry.ETag, grainId, "Update");
+
+                grainState.ETag = NewEtag();
+                grainStateStorage[grainId] = grainState;
+                return grainState.ETag;
             }
 
-            public void DeleteGrainState(string grainStoreKey, string receivedEtag)
+            public string DeleteGrainState(string grainId, string eTag)
             {
-                long currentETag = 0;
-                Tuple<IDictionary<string, object>, long> oldState;
-                if (grainStateStorage.TryGetValue(grainStoreKey, out oldState)){
-                    currentETag = oldState.Item2;
+                IGrainState entry;
+                grainStateStorage.TryGetValue(grainId, out entry);
+                if (entry == null)
+                {
+                    return eTag;
                 }
-                ValidateEtag(currentETag, receivedEtag, grainStoreKey, "Delete");
-                grainStateStorage.Remove(grainStoreKey);
+
+                ValidateEtag(eTag, entry.ETag, grainId, "Delete");
+                grainStateStorage.Remove(grainId);
+                return NewEtag();
+            }
+
+            private static string NewEtag()
+            {
+                return DateTime.UtcNow.ToString(CultureInfo.InvariantCulture);
```

Comment:
that is  not guaranteed to be unique. UtcNow can/will return same time if called close by. Need to ensure different each time. That was among the bugs fixed as part of memory store fixes.